chain/chain/src/flat_storage_resharder.rs

```diff
@@ -62,71 +52,81 @@ use std::iter;
-    sender: ReshardingSender,
     /// Controls cancellation of background processing.
-    pub controller: FlatStorageResharderController,
+    pub handle: ReshardingHandle,
     /// Configuration for resharding.
     resharding_config: MutableConfigValue<ReshardingConfig>,
-    #[cfg(feature = "test_features")]
-    /// TEST ONLY.
-    /// If non zero, the start of scheduled tasks (such as split parent) will be postponed by
-    /// the specified number of blocks.
-    pub adv_task_delay_by_blocks: BlockHeightDelta,
 }
 
 impl FlatStorageResharder {
-    /// Creates a new `FlatStorageResharder`.
-    ///
-    /// # Args:
-    /// * `runtime`: runtime adapter
-    /// * `sender`: component used to schedule the background tasks
-    /// * `controller`: manages the execution of the background tasks
-    /// * `resharding_config`: configuration options
     pub fn new(
+        epoch_manager: Arc<dyn EpochManagerAdapter>,
         runtime: Arc<dyn RuntimeAdapter>,
-        sender: ReshardingSender,
-        controller: FlatStorageResharderController,
+        controller: ReshardingHandle,
         resharding_config: MutableConfigValue<ReshardingConfig>,
     ) -> Self {
-        let resharding_event = Arc::new(Mutex::new(None));
-        Self {
-            runtime,
-            resharding_event,
-            sender,
-            controller,
-            resharding_config,
-            #[cfg(feature = "test_features")]
-            adv_task_delay_by_blocks: 0,
-        }
+        Self { epoch_manager, runtime, handle: controller, resharding_config }
     }
 
-    /// Starts a resharding event.
-    ///
-    /// For now, only splitting a shard is supported.
-    ///
-    /// # Args:
-    /// * `event_type`: the type of resharding event
-    /// * `shard_layout`: the new shard layout
-    pub fn start_resharding(
-        &self,
-        event_type: ReshardingEventType,
-        shard_layout: &ShardLayout,
-    ) -> Result<(), Error> {
-        match event_type {
-            ReshardingEventType::SplitShard(params) => self.split_shard(params, shard_layout),
+    /// Main function to start resharding. This function is a long running cancellable task.
+    /// This is called from the resharding actor and is blocking till the resharding of flat storage
+    /// is completed.
+    pub fn start_resharding(&self, event: &ReshardingSplitShardParams) -> Result<(), Error> {
+        let status = self.runtime.store().flat_store().get_flat_storage_status(event.parent_shard);
+        let Ok(FlatStorageStatus::Ready(FlatStorageReadyStatus { flat_head })) = status else {
+            error!(target: "resharding", ?status, ?event, "flat storage shard split task: parent shard is not ready");
+            panic!("impossible to recover from a flat storage split shard failure!");
+        };
+
+        let next_epoch_id =
+            self.epoch_manager.get_next_epoch_id_from_prev_block(&event.resharding_block.hash)?;
```

Comment:
This is a bit strange, the resharding block is the last of old shard layout. It's used as the prev block so the current block would be the first in the first epoch of new shard layout. Then we go to the next epoch from there and end up in the second epoch with the new shard layout. This is typically fine as we don't schedule two reshardings immediately one after another but it's not technically correct. If what I said is correct please fix (just use the `get_next_epoch_id`).

Reply:
Aah yes you're right! Good catch!